pkg/provider/azure_loadbalancer_backendpool.go

```diff
@@ -889,11 +893,255 @@ func removeNodeIPAddressesFromBackendPool(
+			return true
+		}
+	}
+
+	klog.V(2).Infof("Backend pool %q not found, nothing to remove.", lbBackendPoolName)
+	return false
+}
+
+func (bpi *backendPoolTypePodIP) getAllBackendPoolNamesForEndpointSliceList(service *v1.Service, endpointSliceList []*discovery_v1.EndpointSlice) *utilsets.IgnoreCaseSet {
+	allBPs := utilsets.NewString()
+	ipv4 := bpi.getBackendPoolNamesForEndpointSliceList(service, endpointSliceList, false)
+	ipv6 := bpi.getBackendPoolNamesForEndpointSliceList(service, endpointSliceList, true)
+
+	allBPs.Insert(ipv4.UnsortedList()...)
+	allBPs.Insert(ipv6.UnsortedList()...)
+
+	return allBPs
+}
+
+func (bpi *backendPoolTypePodIP) EnsureHostsInPool(service *v1.Service, _ []*v1.Node, _ string, _ string, _ string, lbName string, backendPool network.BackendAddressPool) error {
+	isIPv6 := isBackendPoolIPv6(pointer.StringDeref(backendPool.Name, ""))
+
+	var (
+		changed               bool
+		numOfAdd, numOfDelete int
+		err                   error
+	)
+
+	endpointSliceList, err := bpi.getEndpointSliceListForService(service)
+
+	if err != nil {
+		klog.Errorf("bpi.EnsureHostsInPool: failed to get endpoint slice list for service %q, error: %s", service.Name, err.Error())
+		return err
+	}
+
+	lbBackendPoolNames := bpi.getBackendPoolNamesForEndpointSliceList(service, endpointSliceList, isIPv6)
+	lbBackendPoolName := pointer.StringDeref(backendPool.Name, "")
+
+	if lbBackendPoolNames.Has(lbBackendPoolName) &&
+		backendPool.BackendAddressPoolPropertiesFormat != nil {
+		if backendPool.LoadBalancerBackendAddresses == nil {
+			lbBackendPoolAddresses := make([]network.LoadBalancerBackendAddress, 0)
+			backendPool.LoadBalancerBackendAddresses = &lbBackendPoolAddresses
+		}
+
+		existingIPs := utilsets.NewString()
+		for _, loadBalancerBackendAddress := range *backendPool.LoadBalancerBackendAddresses {
+			if loadBalancerBackendAddress.LoadBalancerBackendAddressPropertiesFormat != nil &&
+				loadBalancerBackendAddress.IPAddress != nil {
+				klog.V(4).Infof("bpi.EnsureHostsInPool: found existing IP %s in the backend pool %s", pointer.StringDeref(loadBalancerBackendAddress.IPAddress, ""), lbBackendPoolName)
+				existingIPs.Insert(pointer.StringDeref(loadBalancerBackendAddress.IPAddress, ""))
+			}
+		}
+
+		var podIPsToBeAdded []string
+		endpointIPs := utilsets.NewString()
+		var ES *discovery_v1.EndpointSlice
+
+		for _, es := range endpointSliceList {
+
+			if es.AddressType == discovery_v1.AddressTypeIPv6 && !isIPv6 {
+				continue
+			}
+
+			if es.AddressType == discovery_v1.AddressTypeIPv4 && isIPv6 {
+				continue
+			}
+
+			// check if endpoint slice corresponds to given backend pool
+			if bpi.getBackendPoolNameForEndpointSlice(service, es, isIPv6) == lbBackendPoolName {
+				ES = es
+				break
+			}
+		}
+
+		for _, endpoint := range ES.Endpoints {
+
+			if endpoint.Conditions.Ready != nil && !*endpoint.Conditions.Ready {
+				// Skip endpoints that are not ready
+				continue
```

Comment:
do we have logs that we collect from these executions in CCM. If so, can we add additional telemetry to use while debugging.